### PR TITLE
Update examples in C and Fortran

### LINKED ISSUE #236: Simplify examples
- [x] simplify examples `example_c` and `example_f`
- [x] provide a conceptual example in the readme of what spglib does

### PR DESCRIPTION
**Pull request by atztogo**
PR for #236.

- [x] simplify examples `example_c` and `example_f`
- [x] provide a conceptual example in the readme of what spglib does

The examples in C seem not so bad. So I just changed the function names of `test_*` to `example_*`.
The Fortran code of examples in Fortran was very dirty. I made it a little bit cleaner.

Conceptual examples for `spg_get_dataset` should be written in `README.md`.

Closes: #236

### CODE REVIEW

File: example/example.c
Comment (LecrisUT) [inline example/example.c:24]: Could you move them to an `example_full.c` and have a single executable example otherwise for this file?

File: example/example_f08.f90
Comment (LecrisUT) [inline example/example_f08.f90]: Same comment here. Do we need a full list of examples for fortran? We already have the main list in the C one

**Review by LecrisUT — CHANGES_REQUESTED** [review]

File: example/example.c
Comment (atztogo) [reply example/example.c:24]: Is "a single executable example" considered as an example to build an executable? And `example_full.c` would be examples of how to use API's?

File: example/example_f08.f90
Comment (atztogo) [reply example/example_f08.f90]: I don't get many questions about how to use Fortran APIs. If we have a list of C API examples, we may not need many examples for Fortran APIs. A pitfall of Fortran APIs is the different memory orders of arrays. Other important difference is how to use `spg_get_dataset`. In C, it requires to free the memory space of the structure, but in Fortran, `allocatable` arrays are used in the structure, therefore no need to free it explicitly. If these points can be described, the Fortran examples achieve their minimum purpose, I think.

File: example/example.c
Comment (LecrisUT) [reply example/example.c:24]: Yes, a minimal working example, related to the minimal example that will be in the readme

File: example/example_f08.f90
Comment (LecrisUT) [reply example/example_f08.f90]: Fully agree 👍

**Comment by atztogo** [comment]
Since users in C and Fortran may be exclusive, I attempted to change directory structure and the names. 
```
c-example/   # renamed from example/
fortran/example/  # new directory
```
Dependency of `CMakeLists.txt` files should be broken though I don't check it. Since I am afraid of conflict to modify that in upper directories, I just let it mostly unchanged.

**Comment by atztogo** [comment]
It is done except for `CMakeLists.txt` files.

**Comment by LecrisUT** [comment]
> Since users in C and Fortran may be exclusive, I attempted to change directory structure and the names.
> 
> ```
> c-example/   # renamed from example/
> fortran/example/  # new directory
> ```

I agree with that, but can it be structured like:
```
|-- example
  |-- c_api
  |-- fortran_api
  |-- python_api
```

> Dependency of `CMakeLists.txt` files should be broken though I don't check it. Since I am afraid of conflict to modify that in upper directories, I just let it mostly unchanged.

I'll look into that. Just re-request a review to ping me when you want me to look into it.

File: c-example/README.md
Comment (LecrisUT) [inline c-example/README.md]:
```suggestion
This is an example of how to use Spglib through the C api
```

Nitpick, but should we canonicalize `Spglib` or `spglib` or any other form for text names? Only the cmake project name we should stick to the common convention, otherwise, whichever style is more appealing

File: c-example/README.md
Comment (LecrisUT) [inline c-example/README.md]: Can we move this to top level readme? I will suggest some re-wording such that someone clueless like me can understand.

File: c-example/README.md
Comment (LecrisUT) [inline c-example/README.md]: That's what you meant with allocatable in Fortran. I think this needs to be tested in an IDE to confirm this behavior.

Note that using SWIG, we can share C++ smart pointers such that the other language can decide how to deallocate the memory. It has both Python and Fortran bindings.

File: c-example/README.md
Comment (LecrisUT) [inline c-example/README.md]: This line is confusing. It doesn't call a method because it's C not C++, so it does nothing?

File: c-example/example.c
Comment (LecrisUT) [inline c-example/example.c]: Can we add Wikipedia references in the comments to some of these? It's fine only in the C  example because it will be the main entry-point.

File: fortran/example/README.md
Comment (LecrisUT) [inline fortran/example/README.md]: Keep the formatting consistent with the previous one.

File: fortran/example/README.md
Comment (LecrisUT) [inline fortran/example/README.md]: Note-to-self: propose pkg-config flags

File: fortran/example/README.md
Comment (LecrisUT) [inline fortran/example/README.md]: Similarly, Concept should be in top-level, here we specify more language specific notes.

File: fortran/example/README.md
Comment (LecrisUT) [inline fortran/example/README.md]: This can be tested with memcheck or valgrind or...?

**Review by LecrisUT — CHANGES_REQUESTED** [review]

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: I have no clear view on the structure of README.md. Could it be possible for you to manage it after merging this PR?

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: In the Fortran wrapper:

https://github.com/spglib/spglib/blob/02159eef6e7349535049a43fe2272bb634c77945/fortran/spglib_f08.f90#L525

the data in `SpglibDataset` in C structure is copied to the Fortran structure in which the data are allocated by Fortran's allocatable arrays, which should be deallocated automatically as the language specification.

I have no idea about SWIG. For me, writing Python C-API is easy. But for long term perspective, if maintainers can prefer SWIG or any other utils (e.g. cython in python), we can move. Until now, I could be only the person who maintains spglib in long-run, so it has been safer to choose the ways that are easiest for me. However this way is clearly not sustainable although it seems spglib is expected to live longer.

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: This did nothing, but just accessing a structure item. I changed it by 3e038d5, which may be better?

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: I don't see what you mean about wikipedia references...

File: fortran/example/README.md
Comment (atztogo) [reply fortran/example/README.md]: Yes, thanks. Like this d420c5d?

File: c-example/README.md
Comment (LecrisUT) [reply c-example/README.md]: I mean at the `/README.md` instead of `c-example/README.md`

File: fortran/example/README.md
Comment (atztogo) [reply fortran/example/README.md]: I didn't test with memcheck or valgrind, etc. As written above about your comment on `c-example/README.md` (Fortran wrapper uses allocatable arrays).

File: c-example/README.md
Comment (LecrisUT) [reply c-example/README.md]: Indeed, but can you add a comment why it should be that number for example?

File: c-example/example.c
Comment (LecrisUT) [reply c-example/example.c]: Like what's the reference to `International symobl` and so on.

File: fortran/example/README.md
Comment (LecrisUT) [reply fortran/example/README.md]: Yes, but it might be problematic depending on who initialized, e.g. if fortran initializes an instance of the structure `dataset`, but C initializes the pointers inside it, then when fortran destroys the instance, the C initialized variables still exist. Same could happen if we resize arrays and so on. A proper C++ class can automatically avoid those though.

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: If so, we should refer International Tables for Crystallography or Bilbao crystallographic server because wikipedia may be misleading.

File: fortran/example/README.md
Comment (LecrisUT) [reply fortran/example/README.md]: :+1:

File: c-example/README.md
Comment (LecrisUT) [reply c-example/README.md]: > the data in `SpglibDataset` in C structure is copied to the Fortran structure

But that means the original dataset generated by C still needs to be deleted.

> I have no idea about SWIG. For me, writing Python C-API is easy.

Basically SWIG = you don't need to write the binding api, it generates for you. You just point it to the relevant functions/classes and it generates the bindings necessary (including C-Fortran string passing, oh the horror in doing that manually)

File: c-example/example.c
Comment (LecrisUT) [reply c-example/example.c]: Of course, whichever sources can explain it to new-commers.

File: fortran/example/README.md
Comment (LecrisUT) [reply fortran/example/README.md]:
```suggestion
gfortran example.f90 -I[include directory] -L[library directory] -lsymspg
```

With a note that these are taken from `pkg-config --cflags --libs spglib_f08`

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: > But that means the original dataset generated by C still needs to be deleted.

This is done inside the Fortran wrapper.

https://github.com/spglib/spglib/blob/02159eef6e7349535049a43fe2272bb634c77945/fortran/spglib_f08.f90#L669

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: I see.

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: Yes. It is done at 1baad19.

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: OK, it is done at 1baad19.

File: fortran/example/README.md
Comment (atztogo) [reply fortran/example/README.md]: This is properly treated  inside the Fortran wrapper.

https://github.com/spglib/spglib/blob/02159eef6e7349535049a43fe2272bb634c77945/fortran/spglib_f08.f90#L669

File: fortran/example/README.md
Comment (atztogo) [reply fortran/example/README.md]: This is treated at 5800eb0.

File: c-example/README.md
Comment (atztogo) [reply c-example/README.md]: This was treated at 5800eb0,

File: c-example/README.md
Comment (LecrisUT) [reply c-example/README.md]: 👍 Whenever I look at C and Fortran code I miss the automatic memory management of C++

File: c-example/README.md
Comment (LecrisUT) [reply c-example/README.md]: Yes, I like this one. I'll maybe simplify it later and move some back to the C api

File: c-example/example.c
Comment (LecrisUT) [inline c-example/example.c]: What is stored here, recommended allocated size unless we are using safe string commands

File: c-example/example.c
Comment (LecrisUT) [inline c-example/example.c]:
```suggestion
    // SpglibDataset has to be freed after use.
```

File: c-example/example.c
Comment (LecrisUT) [inline c-example/example.c]: Title comment here or above headers. Also short introduction in current folder Readme.

Comment (LecrisUT) [reply c-example/example.c]: Nitpick:

```suggestion
int main(int argc, char* argv[]) {
```

Good to enforce standard practices even in examples.

File: c-example/example.c
Comment (LecrisUT) [inline c-example/example.c]: Move 1e-5 to variable or explain the value

File: c-example/example.c
Comment (LecrisUT) [inline c-example/example.c]:
```suggestion
    printf("International symbol: %s (%d)\n", dataset->international_symbol,
```

File: fortran/README.md
Comment (LecrisUT) [inline fortran/README.md]: Change this to developer's readme, i.e. notes on how to design the binding, what fortran version we use and what type of allocatable arrays we use.

We treat the example in the other readme

File: fortran/example/README.md
Comment (LecrisUT) [inline fortran/example/README.md]: Reference pkg-config command instead of cmake (if they have built it with cmake they would have necessary files to use cmake)

**Review by LecrisUT — CHANGES_REQUESTED** [review]

**Comment by LecrisUT** [comment]
Almost done. One final thing though, why not the following folder structure:

> ```
> |-- example
>   |-- c_api
>   |-- fortran_api
>   |-- python_api
> ```

(/example/c_api, etc.)

Top example readme tells about what common minimal example is implemented in all 3 and that a complete one is also provided.

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: Yes! Done at 4821e31.

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: Thanks! Done at 4821e31.

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: Yes. Done at 4821e31..

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: I agree. Done at 4821e31.

File: c-example/example.c
Comment (atztogo) [reply c-example/example.c]: Done at 4821e31.

File: fortran/example/README.md
Comment (atztogo) [reply fortran/example/README.md]: I have no idea about pkg-config...

File: fortran/README.md
Comment (atztogo) [reply fortran/README.md]: Which do you mean developers of spglib fortran interface or developers who use spglib?

File: fortran/example/README.md
Comment (LecrisUT) [reply fortran/example/README.md]: Add something like:
> Where `[include directory]` and `[library directory]` are taken from `pkg-config --cflags --libs spglib_f08`

You can try manually:
```console
$ PKG_CONFIG_PATH=/path/to/spglib_f08.pc pkg-config --cflags --libs spglib_f08
```

File: fortran/README.md
Comment (LecrisUT) [reply fortran/README.md]: Developers of spglib fortran interface

File: c-example/example.c
Comment (LecrisUT) [reply c-example/example.c]: You can just mark the conversation resolved for these small changes

**Comment by LecrisUT** [comment]
Ok, my final note, fortran files should be `F90` instead of `f90`. It's related to how they read macros and everything. It's a good standard to use.

File: fortran/example/README.md
Comment (atztogo) [reply fortran/example/README.md]: OK, thanks. It is done at c4a2e99.

Comment (atztogo) [reply fortran/example/README.md]: Thanks. It is done at db4b8c0.

File: fortran/README.md
Comment (atztogo) [reply fortran/README.md]: I see. Probably you mean something like this 7cd1c5b.

**Comment by atztogo** [comment]
CI fails because `CMakeLists.txt`s are not updated properly due to the change of directory structure and maybe `f90` to `F90`. I think this is better treated in another PR after merging this PR and expect @LecrisUT's help.

**Comment by LecrisUT** [comment]
Please rebase onto develop. We cannot merge because of merge conflict.

```console
$ git rebase origin/develop
$ git status
Check which files are conflicting and how to resolve it
$ emacs example/c_api/README.md
Command may be different because file was moved
Resolve conflicts by editting the blocks:
<<<<< HEAD
Content from origin/develop
=====
Content from branch examples
>>>>>
$ git rebase --continue
```

I can also recommend gitkraken if you don't feel comfortable with the git cli for this.

I will add commit fixes on top of this PR before merging when you feel it is ready.

File: example/fortran_api/README.md
Comment (LecrisUT) [inline example/fortran_api/README.md]: In the instructions we should not specify `PKG_CONFIG_PATH`. The user should have it installed on the system for these examples. `PKG_CONFIG_PATH` is only for local development and testing.

File: fortran/README.md
Comment (LecrisUT) [reply fortran/README.md]: Well, change the title for starters:
```suggestion
# Fortran interface
```

Also remove `## Conceptual example`. Otherwise that looks ok

File: example/fortran_api/README.md
Comment (atztogo) [reply example/fortran_api/README.md]: Yes, this is done at db4b8c0.

**Comment by atztogo** [comment]
> Please rebase onto develop. We cannot merge because of merge conflict.

Thanks for your detailed instruction. No conflict was found in rebase.

**Comment by LecrisUT** [comment]
Great. Pre-commit seems to fail

Oh, the merge is reporting changes from develop. A rebase would have made it cleaner to review.

**Comment by atztogo** [comment]
> Oh, the merge is reporting changes from develop. A rebase would have made it cleaner to review.

How can I do?

**Comment by LecrisUT** [comment]
> > Oh, the merge is reporting changes from develop. A rebase would have made it cleaner to review.
> 
> How can I do?

At this point will have to interactively rebase to drop the merge commit. But basically running `git rebase` instead of `git merge`.

I'll handle that later

**Comment by atztogo** [comment]
> At this point will have to interactively rebase to drop the merge commit. But basically running `git rebase` instead of `git merge`.

I made `git rebase` but when I wanted to push, git said I need to pull, and I did it with the merge way. Probably my last choice was wrong. What could I do in this case? I should force-push?

**Comment by LecrisUT** [comment]
https://github.com/atztogo/spglib/blob/4d66a9c249bbc96154585a01372ac8e82e41fa96/fortran/CMakeLists.txt#L2

Edit the file extension in the cmake files also.

> I made `git rebase` but when I wanted to push, git said I need to pull, and I did it with the merge way. Probably my last choice was wrong. What could I do in this case? I should force-push?

Yes, you can freely force-push in these situations. Just never force-push the `develop` branch.

**Comment by atztogo** [comment]
> Edit the file extension in the cmake files also.

Yes, I made it. e2a01d5.

**Comment by atztogo** [comment]
Is there anything left to do?

**Comment by LecrisUT** [comment]
Just to fix git history and cmake. I'll do that tomorrow

**Comment by atztogo** [comment]
> Just to fix git history and cmake. I'll do that tomorrow

It is clearly difficult for me to do. Thanks @LecrisUT.

**Review by LecrisUT — APPROVED** [review]
@atztogo I have refactored your branch and fixed the examples. I have noticed that the python example is not equivalent to the others. Could you make some changes there, and if possible add some words to `example/README.md`?

You can work on the same branch again by:
```console
$ git fetch
$ git reset --hard origin/examples
```

Otherwise we can open a new PR for that.

And of course, thanks for all the work on this :smile:

**Review by LecrisUT — APPROVED** [review]
Thank you fully :+1: from me. Do you want to wait for @lan496 to review as well?

**Comment by atztogo** [comment]
> Could you make some changes there, and if possible add some words to `example/README.md`?

Thank you for pointing it out. I did it at ddcc153 and 86566c4.

**Comment by lan496** [comment]
I have no objection. Please merge if you are ready.

**Comment by atztogo** [comment]
Thanks @LecrisUT and @lan496!

**Pull request merged**
